ILRepack.Tests/Utils/BamlDocumentMatcher.cs

```diff
@@ -0,0 +1,154 @@
+    /// current usage which generates a known set of records).
+    /// </summary>
+    internal class BamlDocumentMatcher : Constraint
+    {
+        private readonly BamlDocument _expectedDocument;
+        private object _actualValue, _expectedValue, _context;
+
+        private readonly Dictionary<Type, List<string>> _propertiesToCheck = new Dictionary<Type, List<string>>
+        {
+            { typeof(AssemblyInfoRecord), new List<string> { "AssemblyFullName", "AssemblyId" } },
+            { typeof(AttributeInfoRecord), new List<string> { "AttributeId", "Name", "AttributeUsage", "OwnerTypeId" } },
+            { typeof(ElementStartRecord), new List<string> { "Flags", "TypeId" } },
+            { typeof(XmlnsPropertyRecord), new List<string> { "Prefix", "XmlNamespace", "AssemblyIds" } },
+            { typeof(PropertyListStartRecord), new List<string> { "AttributeId" } },
+            { typeof(PropertyWithConverterRecord), new List<string> { "AttributeId", "ConverterTypeId", "Value" } },
+            //TODO: Too lazy to do reference-only check for this.
+            //   { typeof(DeferableContentStartRecord), new List<string> { "Record" } }
+        };
+
+        public BamlDocumentMatcher(BamlDocument expectedDocument)
+        {
+            _expectedDocument = expectedDocument;
+        }
+
+        public override bool Matches(object actual)
+        {
+            var bamlDocument = (BamlDocument)actual;
+
+            if (!AreEqual(bamlDocument.Signature, _expectedDocument.Signature))
+            {
+                _context = "Document Signature";
+                return false;
+            }
+
+            if (!HaveSameVersion("ReaderVersion", bamlDocument, _expectedDocument) ||
+                !HaveSameVersion("WriterVersion", bamlDocument, _expectedDocument) ||
+                !HaveSameVersion("UpdaterVersion", bamlDocument, _expectedDocument))
+            {
+                return false;
+            }
+
+            return AreRecordsEquivalent(
+                GetRelevantRecords(_expectedDocument),
+                GetRelevantRecords(bamlDocument));
+        }
+
+        private bool HaveSameVersion(
+            string versionProperty, BamlDocument actualDocument, BamlDocument expectedDocument)
+        {
+            var actualVersion = (BamlDocument.BamlVersion)actualDocument.GetPropertyValue(versionProperty);
+            var expectedVersion = (BamlDocument.BamlVersion)expectedDocument.GetPropertyValue(versionProperty);
+
+            _context = versionProperty + ".Major";
+            if (!AreEqual(actualVersion.Major, actualVersion.Major))
+                return false;
+
+            _context = versionProperty + ".Minor";
+            return AreEqual(actualVersion.Minor, expectedVersion.Minor);
+        }
+
+        private bool AreEqual(object expected, object actual)
+        {
+            _expectedValue = expected;
+            _actualValue = actual;
+
+            return Equals(_expectedValue, _actualValue);
```

Comment:
The initial way I did this was with the nice *dynamic* keyword, but we don't get proper error results (e.g.: which property failed).